common/core/flow_engine.c

```diff
@@ -0,0 +1,265 @@
+ * of the rights granted to you under the License to provide to third
+ * parties, for a fee or other consideration (including without
+ * limitation fees for hosting or consulting/ support services related
+ * to the Software), a product or service whose value derives, entirely
+ * or substantially, from the functionality of the Software. Any license
+ * notice or attribution required by the License must also include
+ * this Commons Clause License Condition notice.
+ *
+ * Software: All X1Wallet associated files.
+ * License: MIT
+ * Licensor: HODL TECH PTE LTD
+ *
+ ******************************************************************************
+ */
+
+/*****************************************************************************
+ * INCLUDES
+ *****************************************************************************/
+#include "flow_engine.h"
+
+#include "array_list.h"
+
+/*****************************************************************************
+ * EXTERN VARIABLES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * PRIVATE MACROS AND DEFINES
+ *****************************************************************************/
+#define ENGINE_NUM_BUFFERS 2
+
+#define ENGINE_RUN_INITIALIZE_CB(cb, data_ptr)                                 \
+  {                                                                            \
+    if (cb) {                                                                  \
+      cb(data_ptr);                                                            \
+    }                                                                          \
+  }
+
+#define ENGINE_RUN_EVENT_CB(cb, data_ptr, event)                               \
+  {                                                                            \
+    if (cb) {                                                                  \
+      cb(event, data_ptr);                                                     \
+    }                                                                          \
+  }
+
+/*****************************************************************************
+ * PRIVATE TYPEDEFS
+ *****************************************************************************/
+typedef struct {
+  flow_step_t **buffer;
+  array_list_t *array_list_config;
+} engine_ctx_t;
+
+/*****************************************************************************
+ * STATIC VARIABLES
+ *****************************************************************************/
+static flow_step_t *engine_buffer_0[ENGINE_STACK_DEPTH] = {0};
+static array_list_t engine_list_0;
+
+static flow_step_t *engine_buffer_1[ENGINE_STACK_DEPTH] = {0};
+static array_list_t engine_list_1;
+
+static engine_ctx_t engine_ctx[ENGINE_NUM_BUFFERS] = {
+    {&engine_buffer_0[0], &engine_list_0},
+    {&engine_buffer_1[0], &engine_list_1}};
```

Comment:
Since, the context of an app or flow is maintained by the object of `array_list` and buffer, making it local for the `flow_engine` restricts the usage of `array_list` and relies on an index to control the objects from. This leaves room for incorrect buffer usage when hard-coded indexes are used.

An alternative is to let an application declare its own buffers with its own length etc. The buffers can be passed to the relevant callbacks to manipulate the flow.

Reply:
I have updated the engine APIs to accept buffer from flows and the context is passed on to the callbacks as well..